2261 endpoints for storage and messaging config setup status

adamsachs opened the pull request:
Closes #2261 

https://github.com/ethyca/fidesdocs/issues/56 has been created to track doc updates related to this and other related functionality updates.

### Code Changes

* `GET /messaging/default/status` endpoint that returns a high-level "status" of messaging configuration
    * messaging is considered successfully configured iff the `notifications.notification_service_type` property is set and it points to a messaging service type with a (default) configuration that has all necessary details and secrets populated
* similar `GET /storage/default/status` endpoint that returns a high-level "status" of storage configuration
    * storage is considered successfully configured iff the `storage.active_default_storage_type` property (only settable through API) is set to `s3` and a default s3 storage configuration is present and fully populated with necessary details and secrets (if applicable)
* response to both endpoints looks like the following:
```
{
  "config_status": "configured",
  "detail": "Active default [storage location/messaging service] of type [s3/MAILGUN] is fully configured"
}
```
or
```
{
  "config_status": "not_configured",
  "detail": "No active default [messaging/storage] configuration found"
}
```


### Steps to Confirm

* run your fides server
* try hitting the endpoint(s) in various different states, with messaging/storage fully configured or not
* some useful endpoints to update storage and messaging config states in order to evaluate the endpoints added here:
    * `PATCH /api/v1/config`
    * `GET /api/v1/config?api_set=true`
    * `PUT /api/v1/storage/default`
    * `PUT /api/v1/storage/default/{storage_type}/secret`
    * `GET /api/v1/storage/default/active`
    * `PUT /api/v1/messaging/default`
    * `PUT /api/v1/messaging/default/{service_type}/secret`
    * `GET /api/v1/messaging/default/active`


### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [x] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [x] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [x] Update `CHANGELOG.md`

### Description Of Changes

AC in the issue describes what qualifies as "fully configured" for both messaging and storage -- copied here:
- Complete for storage configuration is true only if S3 configuration (with secrets) is complete AND is set as the active default
- Complete for messaging configuration is true if the active default messaging type has all of its configuration details (secrets and necessary connection details) configured

eastandwestwind commented on src/fides/api/ops/api/v1/endpoints/storage_endpoints.py:
```diff
@@ -331,6 +336,90 @@ def get_active_default_config(
+    # at this point, we only treat the active default storage as fully configured
+    # if it is an s3 storage location. eventually we may need to adjust this when
+    # we support other object stores/storage destinations as valid for production.
+    if storage_config.type != StorageType.s3:
```
could we store the supported active storage types in a const somewhere? Just so it's easier / more obvious to update later on?

adamsachs replied:
yup, great idea

adamsachs replied:
updated in https://github.com/ethyca/fides/commit/f5b36f74bfe0251f1725f6c9b7cc1570e82b2f44, resolving for now but we can create a follow-up if it needs any adjustment

eastandwestwind commented on src/fides/api/ops/api/v1/endpoints/storage_endpoints.py:
```diff
@@ -331,6 +336,90 @@ def get_active_default_config(
+
+    # currently, secrets are only required for s3 storage with auth method `SECRET_KEYS`
+    secrets = storage_config.secrets
+    if (
```
same here, can we make this logic more scalable to handle additional supported types in the future?

adamsachs replied:
hmm, yeah, let me see. for this one, abstracting it out may make it more confusing and not worth the effort at this point, but i'll see what i can come up with.

adamsachs replied:
updated in https://github.com/ethyca/fides/commit/f5b36f74bfe0251f1725f6c9b7cc1570e82b2f44, resolving for now but we can create a follow-up if it needs any adjustment

eastandwestwind reviewed (approved):
Minor comments, none blocking. 

Would also like to see some Postman collection and/or docs updates, but feel free to put in a follow-up ticket if this is urgent.

adamsachs commented on src/fides/api/ops/api/v1/endpoints/messaging_endpoints.py:
```diff
@@ -165,6 +168,76 @@ def get_active_default_config(*, db: Session = Depends(deps.get_db)) -> Messagin
+            service_type=messaging_config.service_type,  # type: ignore
+            secrets=secrets,
+        )
+    except Exception as e:
```
so interestingly when `update_config_secrets` calls `get_schema_for_secrets` it catches both `KeyError` and `ValueError` - but i agree that i don't immediately see how a `KeyError` can be raised. but i figured a general exception handler here is fine since it indicates _something_ is wrong with the secrets, and that's really all we're worried about with this endpoint.

but i know that it's generally not great practice to catch all types of exceptions broadly. i guess i feel like including the `str(e)` can give any additional info about what the specific exception is so that may mitigate things?

now that i think about it though, we should probably be logging that `str(e)` (and wrap it in our `Pii`) and _not_ return it in the endpoint, since it could potentially have sensitive data....

i'll take a closer look at reworking this a bit, thanks for pointing it out!

sanders41 replied:
Just an FYI, I'm fine with whatever decision you all make her. You can catch both `ValueError` and `KeyError` in the same catch: `except (ValueError, KeyError) as e:`

adamsachs replied:
updated in https://github.com/ethyca/fides/pull/2690/commits/f5b36f74bfe0251f1725f6c9b7cc1570e82b2f44, resolving for now but we can create a follow-up if it needs any adjustment

adamsachs commented:
thanks for the quick review @eastandwestwind! i'll look to address your code comments, i think they're worth addressing and the timing here can afford a few iterations.

for docs, i've created this ticket https://github.com/ethyca/fidesdocs/issues/56 and will be addressing all the storage/messaging updates wholesale as part of that.

i'll look to get a postman update in place, but if it's slipping too much then i'll add in a note to cover these endpoints to the follow-up we'd created for the other messaging/storage work!

adamsachs commented:
@eastandwestwind got code updates up addressing your comments, let me know your thoughts on them when you've got a chance.

in the meantime i'll look to update the postman collection!
